include/c2pa.hpp

```diff
@@ -115,4 +125,602 @@ namespace C2pa
         c2pa_release_string(result);
         return;
     }
-}
+
+    // IStream Class wrapper for CStream
+    class CppIStream : public CStream
+    {
+    public:
+        CStream *c_stream;
+
+        template <typename IStream>
+        CppIStream(IStream& istream) {
+            static_assert(std::is_base_of<std::istream, IStream>::value, "Stream must be derived from std::istream");
+            c_stream = c2pa_create_stream(reinterpret_cast<StreamContext*>(&istream), (ReadCallback)reader, (SeekCallback) seeker, (WriteCallback)writer, (FlushCallback)flusher);
+        }
+
+        ~CppIStream() {
+            c2pa_release_stream(c_stream);
+        }
+    private:
+        static size_t reader(StreamContext *context, void *buffer, size_t size) {
```

Comment:
Maybe StreamContext should be passed in as a shared_ptr? It looks like's being. So something like this? 

```cpp
static size_t reader(const std::shared_ptr<StreamContext> context)
```

`std::shared_ptr` is equivalent-ish to an `Rc` in rust. 

Maybe the same thing for void*buffer as well? Using std::unique_ptr or std::shared_ptr documents, via the API, how you expect this value to be cleaned up. IE instead of saying in the docs that the client should delete the *context, the shared_ptr will take care of it after there's no more references.

Reply:
I like the sound of this, but these are callbacks from rust, and I'm not sure things would track. My attempt to make this change resulted in a segfault.